### PR TITLE
feat: show progress output for up command

### LINKED ISSUE #208: Show progress in Listr for slow tasks
Listr can appear to hang when doing a slow operation like downloading images for the first time. We should show this progress.

## Expected Behavior
Progress bar should be included in Listr output

## Current Behavior
Listr appears to hang

## Possible Solution
Redirect stream from docker-compose to Listr task.output?

### PR DESCRIPTION
**Pull request by strophy**
This PR attempts to resolve #208 by piping docker-compose output to the Listr task.

## Issue being fixed or feature implemented
The `start` command calls `docker-compose up` which can be slow if images have not been downloaded yet. Previously no status was displayed, which could leave the user thinking mn-bootstrap was hanging or unresponsive.

## What was done?
Return log output from `docker-compose.upAll` to the listr function which originally called it

## How Has This Been Tested?
Tried it with normal startup and when downloading an image.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by shumkov** [comment]
Maybe it's better to just add build and pull as separate tasks in listr?

**Comment by strophy** [comment]
Picked this up again and tried to refactor by subscribing to the stream while waiting for the promise to complete like we discussed on the call. It seems the promise always completes immediately, even if I return the promise itself and not the `.out` of the promise. I've tried three different implementations (see comments in code) but I'm getting really confused between promises, streams, readables and observables here, what should I be using?

File: src/listr/tasks/startNodeTaskFactory.js [hunk 1]
```diff
@@ -55,9 +52,76 @@ function startNodeTaskFactory(dockerCompose) {
...
+          //   });
+          // }
+          // await waitStream();
+
+          // Doesn't wait
+
+          // ---------------------------------------------------------------------
+
+          // await new Promise((resolve, reject) => {
+          //   const readable = Readable.from(dockerCompose.pull(config.toEnvs()));
```
Comment (shumkov) [inline src/listr/tasks/startNodeTaskFactory.js:75]: You need to add an `await` before `docker-compose.pull`

File: src/listr/tasks/startNodeTaskFactory.js [hunk 2]
```diff
@@ -55,9 +52,76 @@ function startNodeTaskFactory(dockerCompose) {
...
+          //   });
+          // });
+
+          // The "iterable" argument must be an instance of Iterable. Received an instance of Promise
+
+          // ---------------------------------------------------------------------
+
+          // let streamy = '';
+          // const theObserver = new Observable((observer) => {
+          //   streamy = from(dockerCompose.pull(config.toEnvs()));
```
Comment (shumkov) [inline src/listr/tasks/startNodeTaskFactory.js:92]: You forgot `await` as well. The pull method responds with a promise which in turn resolves with the stream.

**Comment by strophy** [comment]
OK, but I'm really not sure what the promise resolves to, I can't check it's type because as soon as it is about to resolve, all the variables disappear in my debugger and it pauses on some weird node async internal that I can't see. If I run it in console, I get one of the following errors, depending on if I return with `.out` or without it:
```
TypeError: Cannot read property 'Symbol(Symbol.asyncIterator)' of undefined
    at logChunks (/home/strophy/Code/mn-bootstrap/src/listr/tasks/startNodeTaskFactory.js:90:39)
    at Task.task (/home/strophy/Code/mn-bootstrap/src/listr/tasks/startNodeTaskFactory.js:97:11)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Task.run (/home/strophy/Code/mn-bootstrap/node_modules/listr2/dist/lib/task.js:165:13)
    at async /home/strophy/Code/mn-bootstrap/node_modules/p-map/index.js:57:22

TypeError: readable is not async iterable
    at logChunks (/home/strophy/Code/mn-bootstrap/src/listr/tasks/startNodeTaskFactory.js:90:39)
    at Task.task (/home/strophy/Code/mn-bootstrap/src/listr/tasks/startNodeTaskFactory.js:97:11)
    at Task.run (/home/strophy/Code/mn-bootstrap/node_modules/listr2/dist/lib/task.js:165:37)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

Which makes me think it isn't really a stream? And yet somehow when I run the code I committed just now, the task waits for `pullAll` to complete properly, but then crashes as soon as it completes like this:

```
strophy@W540:~/Code/mn-bootstrap$ mn start
❯ Start testnet full node
  ✔ Check node is not started
  ⠏ Download updated services
  ◼ Start services  
Pulling core                  ... done
Pulling dapi_insight          ... done
Pulling dapi_tx_filter_stream ... done
Pulling drive_mongodb         ... done
Pulling drive_abci            ... done
Pulling drive_tenderdash      ... done
✔ Start testnet full node 
TypeError: readable is not async iterable
    at logChunks (/home/strophy/Code/mn-bootstrap/src/listr/tasks/startNodeTaskFactory.js:59:39)
    at Task.task (/home/strophy/Code/mn-bootstrap/src/listr/tasks/startNodeTaskFactory.js:66:11)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Task.run (/home/strophy/Code/mn-bootstrap/node_modules/listr2/dist/lib/task.js:165:13)
    at async /home/strophy/Code/mn-bootstrap/node_modules/p-map/index.js:57:22
TypeError: readable is not async iterable
    at logChunks (/home/strophy/Code/mn-bootstrap/src/listr/tasks/startNodeTaskFactory.js:59:39)
    at Task.task (/home/strophy/Code/mn-bootstrap/src/listr/tasks/startNodeTaskFactory.js:66:11)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Task.run (/home/strophy/Code/mn-bootstrap/node_modules/listr2/dist/lib/task.js:165:13)
    at async /home/strophy/Code/mn-bootstrap/node_modules/p-map/index.js:57:22

/home/strophy/Code/mn-bootstrap/node_modules/bls-signatures/blsjs.js:6
var Module=typeof Module!=="undefined"?Module:{};var moduleOverrides={};var key;for(key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}Module["arguments"]=[];Module["thisProgram"]="./this.program";Module["quit"]=(function(status,toThrow){throw toThrow});Module["preRun"]=[];Module["postRun"]=[];var ENVIRONMENT_IS_WEB=false;var ENVIRONMENT_IS_WORKER=false;var ENVIRONMENT_IS_NODE=false;var ENVIRONMENT_IS_SHELL=false;ENVIRONMENT_IS_WEB=typeof window==="object";ENVIRONMENT_IS_WORKER=typeof importScripts==="function";ENVIRONMENT_IS_NODE=typeof process==="object"&&typeof require==="function"&&!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_WORKER;ENVIRONMENT_IS_SHELL=!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_NODE&&!ENVIRONMENT_IS_WORKER;var scriptDirectory="";function locateFile(path){if(Module["locateFile"]){return Module["locateFile"](path,scriptDirectory)}else{return scriptDirectory+path}}if(ENVIRONMENT_IS_NODE){scriptDirectory=__dirname+"/";var nodeFS;var nodePath;Module["read"]=function shell_read(filename,binary){var ret;ret=tryParseAsDataURI(filename);...
```

And I still can't see what is actually happening, because a breakpoint set on line 60 never triggers. How can I debug this?

**Comment by strophy** [comment]
I commented out my function to handle the readable and tried just this:
```
const log = await dockerCompose.pull(config.toEnvs());
console.log(log);
console.log(typeof log);
```

With `await`, the  promise always resolves to `undefined`, so I really don't think it is designed to resolve into a stream. But when I await the function to complete, the output appears correctly (but overwrites listr output). Without `await` we get a pending promise, but it has no key named `.out`. How can I work with data being generated by a promise before it resolves?

I think the log output on stdout is coming from `log: true` in the docker-compose call, i.e. it is writing directly to stdout instead of creating a stream or passing anything back to the calling function. I can't figure out how to convert the output into an async iterable stream that will be accepted as a readable object.

**Pull request closed without merging**